### ISSUE TITLE
Implementing a log in as user...

### ISSUE THREAD

**Issue by bensquire** [bshaffer/oauth2-server-php#573]
Hi all. I've been asked to implement a 'login as user' feature to our project and while I can create a controller that performs something along the lines of:
- Request sent to controller with 'admin access token', and 'desired user'
- Access token confirmed, ensure admin 'client'
- Check requested user id
- Create access token for webapp and desired user (log into DB)
- Return success and access token to admin interface

I was wondering if there was any recommended ways to go abouts this. I realise from a security standpoint this isn't ideal, but we are a small lean startup and this is for now, the simplest solution.

Thanks in advance,
Ben

P.S. As a JS based webapp, we generally use ClientCredentials.

**Comment by bshaffer** [comment]
Client Credentials is not a good option for web-based JS apps, as you are exposing the credentials directly in the client code... unless I misunderstand you, and you're only imbedding the token in the code.

**Comment by bshaffer** [comment]
Sounds like you want to use the JWT Bearer Grant Type. Have you looked into this at all?

**Comment by phindmarsh** [comment]
I have implemented pretty much exactly this, and perhaps the JWT Bearer would have been more appropriate for it (I'd implemented it before that was available in this lib), but this is what I did:

* Create a new Grant Type called `masquerade` that takes an access token and a `user_id` to log in as.
* When administrators are logged into the management system the access token (let's call it `admin_token`) a scope called `masquerade` is attached to it. *This scope can only be requested by admin users*
* To log in as another user, I issue a POST request to the authorise endpoint, sending the `admin_token` and the `user_id` I want to log in as, with the grant type being `masquerade`.
* The custom grant type checks the token is valid, and the scope is correctly associated. If it is, then the grant type returns a new token (say `user_token`) associated with the requested user.
* This `user_token` is swapped into the JS web-app and reloaded. We are then logged in as a given user.

A few things you need to be very careful of:

* Only let this grant type be used by authorised people, either by scope or checking the original user_id is an administrative user. Otherwise anyone will be able to assume the account of anyone else, even admin users. 
* Don't let a `user_token` request a higher scope than the minimum required for you to achieve your login-as requirements (i.e. let someone assume the role of an administrative user).

Hope that helps, if you have any questions feel free to ask :)

**Comment by bensquire** [comment]
Sorry about the slow reply guys, as it happens I finished the initial version of this feature shortly after @bshaffer 2nd comment. Just so we're clear I don't store the users credentials in the JS app, just the access token :)

I haven't looked into JWT Bearer Grant Type, I'll do so now.... thanks :)

In the mean time @phindmarsh would you be able to share you code? I didn't go to the effort of making a new grant-type and basically just created a controller, which does what you described... At the moment we don't actually have any concept of scopes, just users and admin users :S but I take your point, thanks :)

**Comment by phindmarsh** [comment]
@bensquire There isn't much to it, but I've copied the custom grant class and stripped out some of the irrelevant stuff (I extend some abstract grants etc) so it wont 'just work', but it should give you an idea of what to do. Essentially put the logic you have into the `->validateRequest()` method and add the grant when you instantiate your server.

https://gist.github.com/phindmarsh/11c80df8cd220c09b65c

**Comment by dsquier** [comment]
@phindmarsh I ran into a similar issue and tried to document the case and work-around as part of updating the MySQL DDL for this library to the recent 1.7.1 tag, it's over here:

https://github.com/dsquier/oauth2-server-php-mysql

TL;DR Be careful when using scopes with multiple grant types (i.e., Resource Owner Password Credential)
